Doc/library/annotationlib.rst

```diff
@@ -442,93 +442,33 @@ Functions
-remove parameters from the signature, so their annotations are retained:
-
-.. doctest::
-
-   >>> partial_func_kw = partial(func, b="hello")  # Sets default for 'b'
-   >>> get_annotations(partial_func_kw)
-   {'a': <class 'int'>, 'b': <class 'str'>, 'c': <class 'float'>, 'return': <class 'bool'>}
-
-For :class:`functools.partialmethod` objects accessed through a class (unbound),
-the first parameter (usually ``self`` or ``cls``) is preserved, and subsequent
-parameters are handled similarly to :class:`functools.partial`:
-
-.. doctest::
-
-   >>> from functools import partialmethod
-   >>> class MyClass:
-   ...     def method(self, a: int, b: str) -> bool:
-   ...         return True
-   ...     partial_method = partialmethod(method, 1)  # Binds 'a'
-   >>> get_annotations(MyClass.partial_method)
-   {'b': <class 'str'>, 'return': <class 'bool'>}
-
-When a :class:`functools.partialmethod` is accessed through an instance (bound),
-it becomes a :class:`functools.partial` object and is handled accordingly:
-
-.. doctest::
-
-   >>> obj = MyClass()
-   >>> get_annotations(obj.partial_method)  # Same as above, 'self' is also bound
-   {'b': <class 'str'>, 'return': <class 'bool'>}
-
-This behavior ensures that :func:`get_annotations` returns annotations that
-accurately reflect the signature of the partial or partialmethod object, as
-determined by :func:`inspect.signature`.
-
-If :func:`!get_annotations` cannot reliably determine which parameters are bound
-(for example, if :func:`inspect.signature` raises an error), it will raise a
-:exc:`TypeError` rather than returning incorrect annotations. This ensures that
-you either get correct annotations or a clear error, never incorrect annotations:
-
-.. doctest::
-
-   >>> from functools import partial
-   >>> import inspect
-   >>> def func(a: int, b: str) -> bool:
-   ...     return True
-   >>> partial_func = partial(func, 1)
-   >>> # Simulate a case where signature inspection fails
-   >>> original_sig = inspect.signature
-   >>> def broken_signature(obj):
-   ...     if isinstance(obj, partial):
-   ...         raise ValueError("Cannot inspect signature")
-   ...     return original_sig(obj)
-   >>> inspect.signature = broken_signature
-   >>> try:
-   ...     get_annotations(partial_func)
-   ... except TypeError as e:
-   ...     print(f"Got expected error: {e}")
-   ... finally:
-   ...     inspect.signature = original_sig
-   Got expected error: Cannot compute annotations for ...: unable to determine signature
-
-This design prevents the common error of returning annotations that include
-parameters which have already been bound by the partial application.
+Supporting annotations in custom objects
+-------------------------------------------
+
+Objects can support annotation introspection by implementing the :attr:`~object.__annotate__`
+protocol. When an object provides an :attr:`!__annotate__` attribute, :func:`get_annotations`
```

Comment:
More accurately: when an object’s class provides \_\_annotate\_\_, right?

Reply:
Well I was wrong. `__annotate__` is not defined as a special method, but a data descriptor. It’s not looked up on the class.